src/compiler/checker.ts

```diff
@@ -24307,6 +24321,68 @@ namespace ts {
                 /*headMessage*/ undefined,
                 containingMessageChain,
                 errorOutputContainer);
+
+            function checkTagNameDoesNotExpectTooManyArguments(): boolean {
+                const tagType = isJsxOpeningElement(node) || isJsxSelfClosingElement(node) && !isJsxIntrinsicIdentifier(node.tagName) ? checkExpression(node.tagName) : undefined;
+                if (!tagType) {
+                    return true;
+                }
+                const tagCallSignatures = getSignaturesOfType(tagType, SignatureKind.Call);
+                if (!length(tagCallSignatures)) {
+                    return true;
+                }
+                const factory = getJsxFactoryEntity(node);
+                if (!factory) {
+                    return true;
+                }
+                const factorySymbol = resolveEntityName(factory, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ false, node);
+                if (!factorySymbol) {
+                    return true;
+                }
+
+                const factoryType = getTypeOfSymbol(factorySymbol);
+                const callSignatures = getSignaturesOfType(factoryType, SignatureKind.Call);
+                if (!length(callSignatures)) {
+                    return true;
+                }
+
+                let hasFirstparamSignatures = false;
+                // Check that _some_ first parameter expects and SFC-like thing, and that some overload of the SFC expects an acceptable number of arguments
+                for (const sig of callSignatures) {
+                    const firstparam = getTypeAtPosition(sig, 0);
+                    const signaturesOfParam = getSignaturesOfType(firstparam, SignatureKind.Call);
+                    if (!length(signaturesOfParam)) continue;
+                    for (const paramSig of signaturesOfParam) {
+                        hasFirstparamSignatures = true;
+                        if (hasEffectiveRestParameter(paramSig)) {
+                            return true; // some signature has a rest param, so function components can have an aritrary number of arguments
+                        }
+                        const paramCount = getParameterCount(paramSig);
+                        for (const tagSig of tagCallSignatures) {
+                            const tagParamCount = getParameterCount(tagSig);
+                            if (tagParamCount <= paramCount) {

```

Comment:
This arity check is very simple and differs a lot from our other logic. It'll make better errors easier, but are we sure it's okay?

Reply:
I think it's fine; it's pretty conservative. Since we're going through like two calls and then checking _some_ overload is maybe OK (rather than doing overload resolution), and _not_ validating any argument types beyond the first (that's what #29818 does), choosing to only issue an error when we're absolutely sure something's wrong seems reasonable.